fundamental-concepts/contract-keys/4_using_contract_keys.md

```diff
@@ -2,60 +2,88 @@ All Daml update actions that work on contract IDs also have a counter part for c
+  </tr>
+  <tr>
+    <td></td>
+    <td>Contract ID</td>
+    <td>Contract Key</td>
+  </tr>
+    <tr>
+    <td>exercise</td>
+    <td>exercsieCmd</td>
+    <td>exerciseByKeyCmd</td>
+  </tr>
+    <tr>
+    <td>fetch/lookup</td>
+    <td>queryContractId</td>
+    <td>queryContractKey</td>
+  </tr>
+</table>
+<br>
+
+Let's try these out with a Daml script. First, we let `Alice` exercise the `Follow` choice with
+`exerciseByKeyCmd`:
 
 <pre class="file" data-filename="daml/User.daml" data-target="append">
-test = scenario do
-  alice <- getParty "Alice"
-  bob <- getParty "Bob"
-  submit alice $ create User with username=alice, following=[]
-  submit alice $ exerciseByKey @User alice $ Follow bob
+test = script do
+  alice <- allocateParty "Alice"
+  bob <- allocateParty "Bob"
+  submit alice $ createCmd User with username=alice, following=[]
+  aliceContractId <- submit alice $ exerciseByKeyCmd @User alice $ Follow bob
+  return()
 </pre>
 
-- `exerciseByKey` takes as a first argument the template name of the contract we want to exercise a
+- `exerciseByKeyCmd` maps to `exerciseCmd` in Daml and takes as a first argument the template name of the contract we want to exercise a
   choice on preceded by an `@`
 - the second argument is the contract key, in this case `alice`
 
-You can see that `Alice` is now following in the scenario output when you click on `scenario results`:
+You can see that `Alice` is now following in the script output when you click on `script results`:
 
 ![exercise_by_key](assets/exercise_by_key.png)
 
-We can also check this by fetching the contract with `fetchByKey`:
+We can also check this by fetching the contract with `fetchByKey` in Daml. For illustrative purpose we will use `queryContractKey` to show how this done in Daml Script. Before appending the code please delete the `return()` statement.
 
 <pre class="file" data-filename="daml/User.daml" data-target="append">
-  (aliceCid1, aliceUser) <- submit alice $ fetchByKey @User alice
-  assert $ aliceUser == User with username=alice, following=[bob]
-  return ()
-</pre>
 
-- `fetchByKey` takes as first argument the template name, here `@User`, and as a second argument
-  the contract key
-- it returns a tuple, where the first factor is the actual contract ID and the second factor is the
-  contract data
+  userContractIdAndContract <- queryContractKey @User alice (alice)
+  -- We get back an Optional Tuple, like in Haskell we can
+  -- first use fromSome to unpack the Optional Tuple when it is Some Tuple
+  -- then we access the first and second elements with helper functions (fst, snd),
+  -- or alternatively access them positionally with t._1, t._2 and so on
 
-`lookupByKey` is similar to `fetchByKey`, but instead of failing the transaction, it will returns a
-`None` if the given key does not exist.
+  let contractId = fst $ fromSome userContractIdAndContract
+  let contract = snd $ fromSome userContractIdAndContract
```

Comment:
I'd try not to use `fromSome` because it will fail if `userContractidAndContract` is a `None`.

Reply:
How would you extract it then?